### DISCUSSION TITLE
Importing typescript version 3.9.7 code causes compilation error

### DISCUSSION THREAD

**Discussion by heysailor** [vercel/vercel#5202] [General] [answered]
Importing typescript code from a shared folder in a monrepo causes a compilation error on deployment, but not in vercel dev.

[Reproduction is here.](https://github.com/climbingzone/vercel-monorepo)

I have the following structure:

```
projects/name/api/foo
shared/foo
shared/boo
```

The api endpoint `api/foo` is a serverless function which imports values from `shared/boo` and returns them as JSON. One of those values is a string derived in `shared/boo` from `shared/foo`. When called locally with vercel dev, `api/foo` returns correctly as `{"boo":2,"fooString":"3"}`.

When deployed, the following error is returned in the function logs, indicating the file `/var/task/shared/boo/index.js` is not correctly compiled.

```
[GET] /api/foo
12:00:45:90
2020-09-17T04:00:46.455Z	undefined	ERROR	/var/task/shared/boo/index.js:7
const fooed = foo_1.foo.value?.toString();
                              ^
SyntaxError: Unexpected token '.'
    at wrapSafe (internal/modules/cjs/loader.js:1054:16)
    at Module._compile (internal/modules/cjs/loader.js:1102:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/vercel/20d44f83/projects/name/api/foo/index.ts:2:1)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
RequestId: bca9e338-951e-4e63-863c-8cfe43b58400 Error: Runtime exited with error: exit status 1
Runtime.ExitError
```

`Include source files outside of the Root Directory` is selected and the correct root directory is set in the project settings. The error also occurs if I try to import the shared code as modules with yarn workspaces.

It appears this is a bug in compilation of code outside of the project root.... or tell me how to avoid it??

Many thanks.

**Comment by styfle** [comment]
Hi @heysailor 

Node.js doesn't support the optional chaining `?.` syntax yet.

You'll need to add a `tsconfig.json` file with `target: es2018`.

https://www.typescriptlang.org/docs/handbook/compiler-options.html

**Comment by heysailor** [comment]
Ah great, can do. I notice I’m on node 14 locally and Vercel uses 12 which would explain the discrepancy between vercel dev and deployment.

However, seeing as using the optional literal ?. operator is valid typescript and Vercel supports typescript by default, shouldn’t the vercel deployment correctly target es2018 when no tsconfig is supplied, as in this case?

It seems that would be a more optimal, zero-config developer experience.

**Comment by styfle** [comment, marked as answer]
Thanks for the suggestion! I created a bug to track this internally: Story 11466.

We'll probably change the default target from `esnext` to `es2019`.

**Reply by heysailor** [reply]
No problems. Paulo de Mitri is also tracking, via support request #67568
